pkg/operator/datasources.go

```diff
@@ -193,55 +195,75 @@ func (op *Reporting) handlePrometheusMetricsDataSource(logger log.FieldLogger, d
 	// default to importing at the configured import interval
 	importDelay := op.getQueryIntervalForReportDataSource(dataSource)
 
-	var earliestImportedMetricTime, newestImportedMetricTime *metav1.Time
+	var (
+		earliestImportedMetricTime,
+		newestImportedMetricTime,
+		importDataEndTime *metav1.Time
+	)
 	if dataSource.Status.PrometheusMetricImportStatus != nil {
 		if dataSource.Status.PrometheusMetricImportStatus.EarliestImportedMetricTime != nil {
 			earliestImportedMetricTime = dataSource.Status.PrometheusMetricImportStatus.EarliestImportedMetricTime
 		}
 		if dataSource.Status.PrometheusMetricImportStatus.NewestImportedMetricTime != nil {
 			newestImportedMetricTime = dataSource.Status.PrometheusMetricImportStatus.NewestImportedMetricTime
 		}
+		if dataSource.Status.PrometheusMetricImportStatus.ImportDataEndTime != nil {
+			importDataEndTime = dataSource.Status.PrometheusMetricImportStatus.ImportDataEndTime
+		}
 	}
 
 	// determine if we need to adjust our next import and update the status
 	// information if we've imported new metrics.
 	if numResultsImported != 0 {
-		firstTimeRange := results.ProcessedTimeRanges[0]
+		// This is the last timeRange we processed, and we use the End time on
+		// this to determine what time range the importer attempted to import
+		// up until, for tracking our process
 		lastTimeRange := results.ProcessedTimeRanges[len(results.ProcessedTimeRanges)-1]
-		earliestTS := firstTimeRange.Start
+
+		// These are the first and last metric from the import, which we use to
+		// determine the data we've actually imported, versus what we've asked
+		// for.
+		firstMetric := results.Metrics[0]
+		lastMetric := results.Metrics[len(results.Metrics)-1]
 
 		// if there is no existing timestamp then this must be the first import
 		// and we should set the earliestImportedMetricTime
 		if earliestImportedMetricTime == nil {
-			earliestImportedMetricTime = &metav1.Time{earliestTS}
-		} else if earliestImportedMetricTime.After(earliestTS) {
+			earliestImportedMetricTime = &metav1.Time{firstMetric.Timestamp}
+		} else if earliestImportedMetricTime.After(firstMetric.Timestamp) {
 			dataSourceLogger.Errorf("detected time new metric import has older data than previously imported, data is likely duplicated.")
 			// TODO(chance): Look at adding an error to the status.
```

Comment:
if we're redoing statuses in this pull request, have you thought about what to do about this?

Reply:
I'm only really doing changes to Reports right now. There's a lot more to consider with ReportDataSources since we intend to have different kinds in the future, and errors may need to go into conditions, not a specific section of status. The main changes here are what values are being used to set things in the status so that importing progress is tracked more accurately.